### ISSUE TITLE
Dockerfile ADD remote url does not use any HTTP header so always re-downloads

### ISSUE THREAD

**Issue by dkirrane** [docker/docker#15717]
ADD always downloads the file on each build.
```
FROM ubuntu

ADD ["https://oss.sonatype.org/service/local/artifact/maven/content?r=public&g=org.eclipse.xtext&a=org.eclipse.xtext.builder&v=2.9.0-SNAPSHOT", "/opt/xtext-builder.jar"]

CMD ["echo", "hello!"]
```

However, Nexus returns HTTP headers like "Last-Modified" or "ETag" that could be used to prevent the re-download and just use what's in the Docker cache.
```
# curl -I "https://oss.sonatype.org/service/local/artifact/maven/content?r=public&g=org.eclipse.xtext&a=org.eclipse.xtext.builder&v=2.9.0-SNAPSHOT"

HTTP/1.1 200 OK
Content-Disposition: attachment; filename="org.eclipse.xtext.builder-2.9.0-20150820.042448-99.jar"
Content-Length: 341045
Content-Type: application/java-archive
Date: Thu, 20 Aug 2015 10:14:34 GMT
ETag: "{SHA1{44253ea5406c02ead80789cbc763c8c27ba87124}}"
Last-Modified: Thu, 20 Aug 2015 04:24:49 GMT
Server: nginx
Vary: Accept-Charset, Accept-Encoding, Accept-Language, Accept
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Connection: keep-alive
```

**Comment by bfirsh** [comment]
Previously reported in https://github.com/docker/docker/issues/3672 but closed until a better solution could be found. Perhaps we could leave this open to discuss possible solutions to this.

Seems like we track Last-Modified, but it isn't being used to check cache hits: https://github.com/docker/docker/pull/8716

**Comment by aidanhs** [comment]
https://github.com/docker/docker/issues/12361#issuecomment-93992321

> The decision was made to, basically, not trust things like the URL or timestamps, and instead actually check the data itself to make sure nothing has changed.

**Comment by bfirsh** [comment]
Didn't spot that, thanks. ETag seems like a safer way to do it, though. That's typically a hash of the file's contents.

There's probably a certain degree of "good enough" for busting the cache too. E.g. `RUN wget https://...` will never get downloaded again if it changes, and people seem to be alright with that. If you absolutely want a URL to be downloaded again, you can do a `docker build --no cache`.

**Comment by dkirrane** [comment]
Yip I already use RUN `wget` and `build --no-cache` but I'm downloading a bunch of jars and have to manually go an figure out if each jar changed which is cumbersome.

Since `Last-Modified` header will be in a `HTTP-date` format (http://tools.ietf.org/html/rfc2616) couldn't this be parsed & used for validating the cache. If you cannot parse it for any reason revert to the existing approach and re-download the file again. But yes `ETag` would also be another option also.

**Comment by thaJeztah** [comment]
> Since Last-Modified header will be in a HTTP-date format (http://tools.ietf.org/html/rfc2616) couldn't this be parsed & used for validating the cache. If you cannot parse it for any reason revert to the existing approach and re-download the file again. But yes ETag would also be another option also.

Here's an old discussion on that topic; 
see these comments for that; https://github.com/docker/docker/pull/8716#issuecomment-61271301, and https://github.com/docker/docker/pull/8716#issuecomment-61401677

**Comment by aidanhs** [comment]
I am in favour of comparing last-modified/etag as an initial check and falling through to a data-check, was just noting a previous position.

Why bother parsing it? Just hash the last-modified/etag, then compare it to the existing hash. If it matches then great, no need to download.

**Comment by mmyslyvtsev** [comment]
Do we have any follow up on this issue?
Are there any downsides or concerns with using last-modified/etag for caching remote files and avoid re-downloading?

**Comment by thaJeztah** [comment]
I don't think anyone is working on it at the moment; possibly if someone implements @aidanhs's proposal  it could be considered, but I don't know other maintainers opinion on this.

**Comment by LK4D4** [comment]
It's still a problem. However,  I'm not sure if I want to see this implemented in docker. It might be easier to do this by a small bash script.
So, I'm -1 on this.
@thaJeztah I wonder if we have a proper way to close/push forward proposals. Like -3 from maintainers = close, +3 = push forward.

**Comment by efficks** [comment]
It would be great for my build workflow to have this option. Actually, I'm downloading a 200 Mb file everytime!
I don't want to put this file in my git repo! Also, the file could eventually change.

**Comment by alexellis** [comment]
In lieu of a perfect solution being found, can we not review the proposed ideas? @thaJeztah 

It seems like there is a lot of discussion around this and several related issues that have gone nowhere. 

Perhaps the earliest report was over 2 years ago 👎  (in - https://github.com/docker/docker/issues/3672)

**Comment by thaJeztah** [comment]
After a lengthy discussion in our maintainers meeting;

- We're open to adding this in future
- However, the current caching mechanism makes it very complicated to check _either_ on ETAG, _or_ the checksum of the downloaded files, and would require "hacky" changes to the cache-store (which we want to avoid)
- There are plans to refactor the caching store to be more flexible
- After that refactor is done, we can revisit, and see if we can implement this

**Comment by ed-alertedh** [comment]
@thaJeztah Two years later and this still appears to be an issue. I have confirmed the file I'm downloading does have an etag header so I'm assuming this got buried in the backlog...

**Comment by thaJeztah** [comment]
@ed-alertedh I think this is implemented in BuildKit; can you try with BuildKit enabled? (set the `DOCKER_BUILDKIT=1` environment variable before running `docker build`)

**Comment by ed-alertedh** [comment]
Thanks for the speedy response! I wasn't aware of BuildKit, but looks like I can't enable it easily in my setup until this bug is resolved: https://github.com/docker/compose/issues/6440

In the meantime, the file I'm downloading is very unlikely to change at the source so I just used wget instead.

**Comment by ORESoftware** [comment]
For example, with the following Bitbucket urls, it always re-downloads, I assume because Bitbucket fails to include the right header(s). 

```Dockerfile
ADD https://bitbucket.org/teros/datavana/raw/"$commit"/moodwire/mwapi/requirements.txt  '/interos/temp/requirements.txt'
ADD https://bitbucket.org/teros/datavana/raw/"$commit"/moodwire/mwapi/constraints.txt '/interos/temp/constraints.txt'
```

**Comment by thaJeztah** [comment]
@ORESoftware that output looks like you're using the classic builder; can you try using the buildkit builder? See my comment above https://github.com/moby/moby/issues/15717#issuecomment-493854811

**Comment by ORESoftware** [comment]
@thaJeztah thanks - is there documentation on that? link here maybe

**Comment by socketpair** [comment]
> @ed-alertedh I think this is implemented in BuildKit; can you try with BuildKit enabled? (set the `DOCKER_BUILDKIT=1` environment variable before running `docker build`)

Yep. Works for me. It does not re-download. But I did not check if it will if file will be changed.

**Comment by thaJeztah** [comment]
Closing this ticket as I think it's resolved with BuildKit; if there's still cases where re-downloads happen where it's not expected, then it's best to open a new ticket with details in https://github.com/moby/buildkit, which is where the Dockerfile "front-end" code is maintained, which handles this.
